src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, nothing)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)  # LHS is just documentation here
+        # Don't need to store every `c`, need last + one for the next iteration
+    end
+    y = first(last(hobbits))
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
```

Comment:
What about arrays of arrays?

e.g. 
```
julia> foldl((x,y)->y, [[1f0,2f0],[10f0,20f0],[100f0,200f0]])
2-element Vector{Float32}:
 100.0
 200.0
```
which if I was pullingback say `[1.0, 2.0]` that would miss out on ProjectTo fixing types.

Reply:
Sure, this should be widened.

Reply:
There are quite a lot of places we have this idiom. I do wonder if we should have `ProjectTo(_) = identity` or something, in the end.

Reply:
Yeah, we do have an issue discussing that somewhere.
Maybe we should resolve that.